Add generic slice support for 'in' operator.

echlebek opened the pull request:
This commit makes the 'IN' operator more tolerant of input
datatypes; previously all 'IN' statements required []interface{}
as an input.

Now any kind of slice can be used.

Signed-off-by: Eric Chlebek <eric@sensu.io>

Commit:
Add generic slice support for 'in' operator.

This commit makes the 'IN' operator more tolerant of input
datatypes; previously all 'IN' statements required []interface{}
as an input.

Now any kind of slice can be used.

Signed-off-by: Eric Chlebek <eric@sensu.io>

jamesdphillips commented on evaluationStage.go:
```diff
@@ -419,15 +434,45 @@ func separatorStage(left interface{}, right interface{}, parameters Parameters)
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return float64(v.Uint())
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return float64(v.Int())
```
Do we run the risk of overflows causing false positives/negatives when an int64 / uint64 value is nearing is upper bounds?

echlebek replied:
There is some risk here, but not of overflow. Rather, we risk losing information for numbers with a magnitude greater than 53 bits, possibly leading to bad comparisons.

I'd consider this risk to be quite small, and worth it for the sake of simplicity.

jamesdphillips commented on evaluationStage.go:
```diff
@@ -17,6 +17,21 @@ const (
+	reflect.Uint64:  struct{}{},
+	reflect.Float32: struct{}{},
+	reflect.Float64: struct{}{},
+}
```
Entirely to satisfy my own curiosity: why use `struct{}` instead of say `bool`? Convention? Micro-optimization?

jamesdphillips reviewed (approved):
One question but otherwise code looks good to me.

Might be nice to add a benchmark so we can get a sense of the cost of this addition and to give us a baseline for any future optimizations.

mercul3s reviewed (approved):
![](https://media2.giphy.com/media/xUPJUCEj5Z8nz9M9mo/giphy.gif)

Commit:
Add benchmark for in operator.

Signed-off-by: Eric Chlebek <eric@sensu.io>

Commit:
Add benchmark, and optimize the evaluation of IN.

BenchmarkInArrayInteger-4           351 ns/op
BenchmarkInArrayIntegerAndFloat-4   338 ns/op

Signed-off-by: Eric Chlebek <eric@sensu.io>

echlebek commented:
@jamesdphillips added a benchmark as per your request. As you can see, the operation is rather cheap. It shouldn't be a concern compared to other costs in the pipeline.

(See commit message in https://github.com/sensu/govaluate/pull/2/commits/11ae95ab19aa83d8f9950433a9c55d8cf75fe64b)
